### ISSUE TITLE
Shared object corruption to due 32bit roundup()

### ISSUE THREAD

**Issue by eytanaim** [NixOS/patchelf#224]
**Describe the bug**

Using this tool on /lib64/libgcc_s.so.1 (or many other system libraries) on RedHat 5, resulted corruption to the shared object.

After investigating the issue, I believe I have found the root-cause:
**The roundup function, handles int values (32bit) while it is sometimes used with 64bit values:**

```
void ElfFile<ElfFileParamNames>::rewriteSectionsLibrary()
{
    /* For dynamic libraries, we just place the replacement sections
       at the end of the file.  They're mapped into memory by a
       PT_LOAD segment located directly after the last virtual address
       page of other segments. */
    Elf_Addr startPage = 0;
    for (unsigned int i = 0; i < phdrs.size(); ++i) {
        Elf_Addr thisPage = roundUp(rdi(phdrs[i].p_vaddr) + rdi(phdrs[i].p_memsz), getPageSize());
        if (thisPage > startPage) startPage = thisPage;
    }
...
```
When `phdrs[i].p_vaddr` requires 64 bit representation, it leads to the corruption.

**Steps To Reproduce**

* Extract my files [myElfs.zip](https://github.com/NixOS/patchelf/files/4982114/myElfs.zip)
* Run `patchelf --debug --add-needed eytan.so  ./libgcc_s.so.1`
* File is now corrupted even if it finds eytans.so :)
* Output of following commands fails (needless to say libgcc is a dynamic executable):
```
[root@agentcos5smp3 wip]# ldd  libgcc_s.so.1
        not a dynamic executable
[root@agentcos5smp3 wip]# readelf -a libgcc_s.so.1 > /dev/null
readelf: Warning: Virtual address 0x377ac68168 not located in any PT_LOAD segment.
readelf: Error: Unable to read in dynamic data
```
Comparing the readelf output of before and after shows the issue:
The 64bit value of 0x000000377ae***** is being trimmed to 0x000000007ae***** due to a bad roundup
![image](https://user-images.githubusercontent.com/9480886/88544419-06d04780-d022-11ea-8c20-c54450f4025e.png)

**Expected behavior**
File to valid

**`patchelf --version` output**
patchelf 0.11

**Additional context**

**Comment by DerDakon** [comment]
This should help:

```diff
--- a/src/patchelf.cc
+++ b/src/patchelf.cc
@@ -541,7 +541,7 @@ static void writeFile(std::string fileName, FileContents contents)
 }
 
 
-static unsigned int roundUp(unsigned int n, unsigned int m)
+static uint64_t roundUp(uint64_t n, uint64_t m)
 {
     return ((n - 1) / m + 1) * m;
 }
```

**Comment by eytanaim** [comment]
Any reason why this patch is not applied into master?
When fixing this issue I found 2 additional things. I'll update once I find out the RCA.

**Comment by eytanaim** [comment]
After using the roundup for 64 bit I found out 2 additional issues that relates to the same issue  - the fact that the offset value is not equal to the address value for the section and program headers in the files I have attached above.

The following patch fixes both of them. I'm not sure if there are others issues that might raise using the library above.

[patch.txt](https://github.com/NixOS/patchelf/files/5257883/patch.txt)
